### PR TITLE
SMT solidity implementation

### LINKED ISSUE #97: Implement an SMT on chain
zk-kit doesn't include an SMT contract. We may need to write our own implementation.

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #97 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/contracts/contracts/SparseMerkleTree.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        require(leaf < SNARK_SCALAR_FIELD);
+        require(index < 2**depth);
+
+        uint256 hash = leaf;
+        uint256 lastLeftElement;
+        uint256 lastRightElement;
+
+        for (uint8 i = 0; i < depth; ) {
+            self.leaves[i][index] = hash;
+            if (index % 2 == 0) {
```
Comment (vuvoth) [inline packages/contracts/contracts/SparseMerkleTree.sol:58]: I think `index & 1` will more efficiency

File: packages/contracts/contracts/SparseMerkleTree.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    ) public {
+        uint256 depth = self.depth;
+        require(leaf < SNARK_SCALAR_FIELD);
+        require(index < 2**depth);
+
+        uint256 hash = leaf;
+        uint256 lastLeftElement;
+        uint256 lastRightElement;
+
+        for (uint8 i = 0; i < depth; ) {
```
Comment (vuvoth) [inline packages/contracts/contracts/SparseMerkleTree.sol:56]: why we don't uint256? In my view use u256 will make program more sense.

File: packages/contracts/test/SMT.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        poseidon
+    ) as any) {
+        const f = new ethers.ContractFactory(abi, bytecode, accounts[0])
+        const c = await f.deploy()
+        await c.deployed()
+        libraries[`Poseidon${inputCount}`] = c.address
+    }
+    delete libraries['Poseidon3']
+    delete libraries['Poseidon5']
+    const merkleTreeLibFactory = await ethers.getContractFactory(
```
Comment (vuvoth) [inline packages/contracts/test/SMT.ts:19]: Could we move deploy SMT lib to beforeEach?

File: packages/contracts/contracts/SparseMerkleTree.sol (hunk 2 above)
Comment (vimwitch) [reply packages/contracts/contracts/SparseMerkleTree.sol:56]: I think the 8 bit integer is cheaper on the stack. The depth can't be more than 32 so it we never need more than 8 bits here.

File: packages/contracts/test/SMT.ts (hunk 3 above)
Comment (vimwitch) [reply packages/contracts/test/SMT.ts:19]: Maybe, this is the same problem as before. There's no way to pass variables from `beforeEach` to the test function. I'd rather use this function to deploy than have shared state variables being mutated.

File: packages/contracts/contracts/SparseMerkleTree.sol [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    uint256 depth;
+    uint256 root;
+    // depth to zero node
+    mapping(uint256 => uint256) zeroes;
+    // depth to index to leaf
+    mapping(uint256 => mapping(uint256 => uint256)) leaves;
+}
+
+library SparseMerkleTree {
+    uint8 internal constant MAX_DEPTH = 32;
```
Comment (vivianjeng) [inline packages/contracts/contracts/SparseMerkleTree.sol:16]: Can it also be built with a template?
and `64` is better if possible.

**Pull request merged**